### PR TITLE
New AB lib integration

### PR DESCRIPTION
**Pull request by nicl**
# We are removing Flow from Frontend, we should refactor this in after that work is complete

## What does this change?

Use new [ab-testing](https://github.com/guardian/ab-testing/) library for testing code. As part of this, we've also refactored a bit to inline some code and remove things too - in particular, the abillity to force/exclude a test via local storage as the library does not suppor this.

TODOs:

* fix `ab-ui.js` now that it can't use local storage
* check Contributions are happy
* re-add/replace tests from `ab.spec.js`, specifically around # opt-in/out and test ordering
* provide real flow types from ab-testing library (if possible)

And for separate PRs:

* replace Ophan event code with new library as well
* remove Automat-specific code which is no longer necessary

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## What is the value of this and can you measure success?

Behaviour should be unchanged (other than local storage) and tests work as they currently do.

### Tested

- [x] Locally
- [ ] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/experiments/ab.js [hunk 1]
```diff
@@ -2,24 +2,12 @@
...
+import { runnableTestsToParticipations } from 'common/modules/experiments/ab-utils';
 import {
     registerCompleteEvents,
     registerImpressionEvents,
     trackABTests,
 } from 'common/modules/experiments/ab-ophan';
-import {
-    getParticipationsFromLocalStorage,
-    setParticipationsInLocalStorage,
-} from 'common/modules/experiments/ab-local-storage';
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/experiments/ab.js:21]: Is there no way to keep a user in a test group across pageviews?

File: static/src/javascripts/projects/common/modules/experiments/ab-ui.js [hunk 2]
```diff
@@ -1,4 +1,5 @@
 // @flow
+
 import template from 'lodash/template';
 import { local as storage } from 'lib/storage';
 import $ from 'lib/$';
```
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/experiments/ab-ui.js:5]: Unrelated, but: a blast from the past!

File: static/src/javascripts/projects/common/modules/experiments/ab.js (hunk 1 above)
Comment (nicl) [reply static/src/javascripts/projects/common/modules/experiments/ab.js:21]: On CODE/PROD you can do this using the `opt/in` and `opt/out` endpoints: https://github.com/guardian/frontend/blob/master/applications/conf/routes#L61.

But for local development, the only way here would be by modifying your MVT ID. See https://github.com/guardian/ab-testing/#mvtid-calculator for a bit more on how to calculate this.

Comment (nicl) [reply static/src/javascripts/projects/common/modules/experiments/ab.js:21]: I should add, I'm not at all against local storage for this or anything. But we should build this into the `ab-testing` library if we want it.

File: static/src/javascripts/projects/common/modules/experiments/ab-ui.js (hunk 2 above)
Comment (nicl) [reply static/src/javascripts/projects/common/modules/experiments/ab-ui.js:5]: Ha, which bit!? The lodash template, jquery?

File: static/src/javascripts/projects/common/modules/experiments/ab.js (hunk 1 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/experiments/ab.js:21]: Thanks for clarifying.

I’ve used the MVT ID calculator before and it’s very convenient! Better than going down the route of changing your request headers, which I’ve also done…

File: static/src/javascripts/projects/common/modules/experiments/ab-ui.js (hunk 2 above)
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/experiments/ab-ui.js:5]: I’m talking about the jQuery `$` alias. That dollar key got smashed quite a lot—especially if you were writing PHP on the back end, but that would have been _so uncool._

**Comment by jorgeazevedo** [comment]
Hi!

The `master` branch in this repository has been renamed to `main`. We’ve updated the base branch of this PR for you. So if your PR has already been approved, you can just merge as usual.

If you try to push new commits to this branch however, it might not go through due to pre-push hook failures. You’ll need to either rebase this branch on top of `main` or merge from `main` to update your hooks.

Thanks! If you have any other problems feel free to reach out on the dotcom channel.

**Comment by gtrufitt** [comment]
Follow up https://github.com/guardian/frontend/pull/22984/files

File: common/app/conf/switches/ABTestSwitches.scala [hunk 3]
```diff
@@ -101,7 +101,7 @@ trait ABTestSwitches {
     "ab-remote-epic-variants",
     "Serve epics from remote service for subset of audience",
     owners = Seq(Owner.withGithub("nicl")),
-    safeState = Off,
+    safeState = On,
```
Comment (gtrufitt) [inline common/app/conf/switches/ABTestSwitches.scala:104]:
```suggestion
    safeState = Off,
```

File: static/src/javascripts/bootstraps/enhanced/main.js [hunk 4]
```diff
@@ -357,7 +357,7 @@ const bootEnhanced = (): void => {
                     require => {
                         bootstrapContext(
                             'experiments',
-                            require('common/modules/experiments')
+                            require('common/modules/experiments/ab-ui')
```
Comment (gtrufitt) [inline static/src/javascripts/bootstraps/enhanced/main.js:360]: Yeah this makes sense actually, none of the others are index pages

File: static/src/javascripts/projects/common/modules/experiments/ab.js [hunk 5]
```diff
@@ -58,6 +48,9 @@ const buildKeywordTags = page => {
     }));
 };
 
+/* eslint-disable import/no-mutable-exports */
+export const ABLib: ABType = AB(testConfig(concurrentTests));
```
Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/experiments/ab.js:52]: I guess we want this to be a `let` rather than `const`

**Comment by gtrufitt** [comment]
Still relevant

**Comment by nicl** [comment]
Closing as been so long, no longer on the team, conflicts etc.

**Pull request closed without merging**